server/hostid/hostid.go

```diff
@@ -0,0 +1,130 @@
+package hostid
+
+import (
+	"flag"
+	"io"
+	"os"
+	"path"
+	"sync"
+
+	"github.com/buildbuddy-io/buildbuddy/server/util/disk"
+	guuid "github.com/google/uuid"
+)
+
+const (
+	// Name of the configuration directory in os.UserConfigDir
+	configDirName  = "buildbuddy"
+	hostIDFilename = "host_id"
+)
+
+var (
+	hostID      string
+	hostIDError error
+	hostIDOnce  sync.Once
+
+	failsafeID     string
+	failsafeIDOnce sync.Once
+
+	manualHostID = flag.String("executor.host_id", "", "Optional: Allows for manual specification of an executor's host id. If not set, a random UUID will be used.")
+)
+
+func configDir() (string, error) {
+	// HOME and XDG_CONFIG_HOME may not be defined when running with `bazel test`.
+	if testTmpDir := os.Getenv("TEST_TMPDIR"); testTmpDir != "" {
+		return os.MkdirTemp(testTmpDir, "buildbuddy-config-*")
+	}
+
+	userConfigDir, err := os.UserConfigDir()
+	if err != nil {
+		// Home dir is not defined
+		return "", err
+	}
+	configDirPath := path.Join(userConfigDir, configDirName)
+	err = os.MkdirAll(configDirPath, 0755)
+	if err != nil {
+		return "", err
+	}
+	return configDirPath, nil
+}
+
+func getOrCreateHostId(dir string) (string, error) {
+	if *manualHostID != "" {
+		return *manualHostID, nil
+	}
+	if dir == "" {
+		d, err := configDir()
+		if err != nil {
+			return "", err
+		}
+		dir = d
+	} else {
+		// migrate the host_id, if possible
+		// TODO (zoey): remove this after we expect most migrations to be done
+		if oldConfigDir, err := configDir(); err == nil {
+			oldHostIDFilepath := path.Join(oldConfigDir, hostIDFilename)
+			newHostIDFilepath := path.Join(dir, hostIDFilename)
+			// This call will move an old-style host id file (if it exists) to the new
+			// host id file location.
+			_ = disk.MoveFile(oldHostIDFilepath, newHostIDFilepath)
```

Comment:
sorry if my last comment was not 100% clear, I think we should make `disk.copyFileViaTempSibling` public and use that here instead. otherwise LGTM

Reply:
kk, will do

Reply:
done.